File: client/lib/shared/http/models/answer_model.dart
import 'package:client/shared/http/models/vote_model.dart';

import 'author_model.dart';

class AnswerModel {
  final int id;
  final String content;
  final List<VoteModel>? votes;
  final AuthorModel author;
  final int commentsCount;
  final DateTime createdAt;
  final DateTime updatedAt;

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> json = <String, dynamic>{};
    json['id'] = id;
    json['author'] = author.toJson();
    json['content'] = content;
    json['commentsCount'] = commentsCount;
    json['createdAt'] = createdAt.toIso8601String();
    json['updatedAt'] = updatedAt.toIso8601String();
    json['votes'] = votes?.map((e) => e.toJson());

    return json;
  }

  AnswerModel.fromJson(Map<String, dynamic> json)
      : id = json['id'],
        votes = json['votes']
            ?.map<VoteModel>((e) => VoteModel.fromJson(e))
            .toList(),
        commentsCount = json['commentsCount'],
        author = AuthorModel.fromJson(json['author']),
        content = json['content'],
        createdAt = DateTime.parse(json['createdAt']),
        updatedAt = DateTime.parse(json['updatedAt']);
}

File: server/src/interfaces/AnswerModel.ts
import { AuthorModel } from "./AuthorModel";
import { VoteModel, voteModelQuery } from "./VoteModel";

export const answerModelPrismaQuery = {
  id: true,
  content: true,
  votes: {
    select: voteModelQuery,
  },
  author: {
    select: {
      id: true,
      name: true,
      picture: true,
    },
  },
  _count: {
    select: {
      comments: true,
    },
  },
  createdAt: true,
  updatedAt: true,
};

export function convertQueryToAnswerModel(answer: any): AnswerModel {
  console.log(
    "Cheguei aqui. ",
    answer.id,
    answer.content,
    answer.votes,
    answer.author,
    answer.commentsCount,
    answer.createdAt,
    answer.updatedAt
  );
  return <AnswerModel>{
    id: answer.id,
    content: answer.content,
    votes: answer.votes,
    author: answer.author,
    commentsCount: answer._count.comments,
    createdAt: answer.createdAt,
    updatedAt: answer.updatedAt,
  };
}

export interface AnswerModel {
  id: number;
  content: string;
  votes: VoteModel[];
  commentsCount: number;
  author: AuthorModel;
  createdAt: Date;
  updatedAt: Date;
}
